### PR TITLE
Added customMessageType to Publish, Signal, Subscribe, History, File.

### PR DESCRIPTION
**Pull request by marcin-cebo**
feat: Add customMessageType support for the following APIs: publish, signal, share file, subscribe, and history

### CODE REVIEW

**Review by wkal-pubnub — CHANGES_REQUESTED** [review]
Additionally there are some build errors related to missing parameters in JS and iOS in fetchMessages

File: pubnub-gson/pubnub-gson-impl/src/main/java/com/pubnub/internal/java/PubNubForJavaImpl.kt [hunk 1]
```diff
@@ -218,8 +219,8 @@ open class PubNubForJavaImpl(configuration: PNConfiguration) :
         return SignalImpl(this)
     }
 
-    override fun signal(message: Any, channel: String): com.pubnub.api.endpoints.pubsub.Signal {
-        return super.signal(channel, message)
+    override fun signal(message: Any, channel: String): SignalBuilder {
+        return SignalImpl(this).message(message).channel(channel) as SignalBuilder
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply pubnub-gson/pubnub-gson-impl/src/main/java/com/pubnub/internal/java/PubNubForJavaImpl.kt:223]: Good catch.

File: pubnub-kotlin/pubnub-kotlin-impl/src/integrationTest/kotlin/com/pubnub/api/integration/PublishIntegrationTests.kt [hunk 2]
```diff
@@ -840,4 +852,41 @@ class PublishIntegrationTests : BaseIntegrationTest() {
...
+            }
+        })
+
+        subscription.subscribe()
+        assertTrue(connected.await(10, TimeUnit.SECONDS))
+
+        channel.publish(
+            message = generatePayload(),
+            customMessageType = expectedCustomMessageType
+        ).await { resutl: Result<PNPublishResult> ->
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply pubnub-kotlin/pubnub-kotlin-impl/src/integrationTest/kotlin/com/pubnub/api/integration/PublishIntegrationTests.kt:884]: ok, I will change it.

File: pubnub-kotlin/pubnub-kotlin-impl/src/main/kotlin/com/pubnub/internal/endpoints/pubsub/PublishEndpoint.kt [hunk 3]
```diff
@@ -93,6 +98,8 @@ class PublishEndpoint internal constructor(
             queryParams["norep"] = true.valueString
         }
 
+        customMessageType?.run { queryParams[CUSTOM_MESSAGE_TYPE_QUERY_PARAM] = this }
```
[the remark this answers is not in the document]
Comment (marcin-cebo) [reply pubnub-kotlin/pubnub-kotlin-impl/src/main/kotlin/com/pubnub/internal/endpoints/pubsub/PublishEndpoint.kt:101]: I would also use .let but I wanted to be in line with approach used in this method where there are:
```
meta?.run { queryParams["meta"] = pubnub.mapper.toJson(this) }
shouldStore?.run { queryParams["store"] = this.numericString }
ttl?.run { queryParams["ttl"] = this.toString() }
```
I will refactor this method to use:
` .?.let {}`

File: pubnub-gson/pubnub-gson-impl/src/main/java/com/pubnub/internal/java/PubNubForJavaImpl.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (wkal-pubnub) [reply pubnub-gson/pubnub-gson-impl/src/main/java/com/pubnub/internal/java/PubNubForJavaImpl.kt:223]: you're right, Lombok messes with the return type of those methods.
The solution would be to add

```
    public SignalImpl message(Object message) {
        this.message = message;
        return this;
    }

    public SignalImpl channel(String channel) {
        this.channel = channel;
        return this;
    }

    public SignalImpl customMessageType(String customMessageType) {
        this.customMessageType = customMessageType;
        return this;
    }
```

to SignalImpl and disable lombok generating setters... but maybe let's leave the cast here for now, seems like it's simpler even though I don't like it

File: pubnub-kotlin/pubnub-kotlin-core-api/src/commonMain/kotlin/com/pubnub/api/models/consumer/pubsub/files/PNFileEventResult.kt [hunk 4]
```diff
@@ -45,6 +45,9 @@ class PNFileEventResult(
         if (file != other.file) {
             return false
         }
+        if (customMessageType != other.customMessageType) {
```
Comment (wkal-pubnub) [inline pubnub-kotlin/pubnub-kotlin-core-api/src/commonMain/kotlin/com/pubnub/api/models/consumer/pubsub/files/PNFileEventResult.kt:48]: if you modify equals() you MUST modify hashCode() as well to match

Comment (marcin-cebo) [reply pubnub-kotlin/pubnub-kotlin-core-api/src/commonMain/kotlin/com/pubnub/api/models/consumer/pubsub/files/PNFileEventResult.kt:48]: Good catch. I will fix that.

**Review by wkal-pubnub — APPROVED** [review]

**Comment by marcin-cebo** [comment]
@pubnub-release-bot release kotlin as v10.2.0

**Pull request merged**